### PR TITLE
add FIRE minimizer

### PR DESCRIPTION
**Pull request by psn417**

### CODE REVIEW

File: doc/gpumd/input_parameters/minimize.rst [hunk 1]
```diff
@@ -6,18 +6,18 @@
 ================
 
 This keyword is used to minimize the energy of the system.
-Currently, only the steepest descent method has been implemented.
+Currently, the fast inertial relaxation engine (FIRE) [Bitzek2006]_ [Guénolé2020]_ method and the steepest descent (SD) method has been implemented.
```
Comment (brucefan1983) [inline doc/gpumd/input_parameters/minimize.rst:9]: has -> have

File: doc/bibliography.rst [hunk 2]
```diff
@@ -170,3 +170,15 @@ Bibliography
    | J. F. Ziegler, J. P. Biersack, and U. Littmark
    | In *The Stopping and Range of Ions in Matter*, volume 1
    | New York, 1985. Pergamon. ISBN 0-08-022053-3
+
```
Comment (brucefan1983) [inline doc/bibliography.rst:173]: Check if the references here are in alphabetic order. If so, follow the convention.

File: doc/gpumd/input_parameters/minimize.rst [hunk 3]
```diff
@@ -41,6 +41,14 @@ If you have no idea how small :attr:`force_tolerance` should be, you can simply
 
 In this case, the energy minimization process will definitely run 10,000 steps.
 
+Example 3
+^^^^^^^^^
+The command::
+
+  minimize fire 1.0e-5 1000
+
+means that one wants to do an energy minimization using the steepest descent method, with a force tolerance of :math:`10^{-5}` eV/Å for up to 1,000 steps.
```
Comment (brucefan1983) [inline doc/gpumd/input_parameters/minimize.rst:50]: steepest descent -> FIRE

File: src/minimize/minimize.cu [hunk 4]
```diff
@@ -54,6 +55,23 @@ void Minimize::parse_minimize(
...
+      PRINT_INPUT_ERROR("Number of steps should be an integer.");
+    }
+    if (number_of_steps <= 0) {
+      PRINT_INPUT_ERROR("Number of steps should > 0.");
+    }
+  } else if (strcmp(param[1], "fire") == 0) {
+    minimizer_type = 1;
+
+    if (num_param != 4) {
+      PRINT_INPUT_ERROR("minimize sd should have 2 parameters.");
```
Comment (brucefan1983) [inline src/minimize/minimize.cu:68]: minimize sd -> minimize fire

File: src/minimize/minimizer_fire.cu [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+      }
+      N_neg++;
+    } else {
+      next_dt = dt * f_dec;
+      if (next_dt > dt_min)
+        dt = next_dt;
+      alpha = alpha_start;
+      // move position back
+      scalar_multiply(-0.5 * dt, v, temp1);
+      CHECK(cudaDeviceSynchronize());
```
Comment (brucefan1983) [inline src/minimize/minimizer_fire.cu:78]: do not need `cudaDeviceSynchronize`

File: src/minimize/minimizer_fire.cu [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+      vector_sum(position_per_atom, temp1, position_per_atom);
+      CHECK(cudaDeviceSynchronize());
```
Comment (brucefan1983) [inline src/minimize/minimizer_fire.cu:80]: do not need `cudaDeviceSynchronize`

File: src/minimize/minimizer_fire.cu [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+      N_neg = 0;
+    }
+
+    // md step
+    // implicit Euler integration
+    double F_modulus = sqrt(dot(force_per_atom, force_per_atom));
+    double v_modulus = sqrt(dot(v, v));
+    // dv = F/m*dt
+    scalar_multiply(dt / m, force_per_atom, temp2);
+    CHECK(cudaDeviceSynchronize());
```
Comment (brucefan1983) [inline src/minimize/minimizer_fire.cu:91]: do not need `cudaDeviceSynchronize`

File: src/minimize/minimizer_fire.cu [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+    vector_sum(v, temp2, v);
+    CHECK(cudaDeviceSynchronize());
```
Comment (brucefan1983) [inline src/minimize/minimizer_fire.cu:93]: do not need `cudaDeviceSynchronize`

File: src/minimize/minimizer_fire.cu [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+    scalar_multiply(1 - alpha, v, temp1);
+    scalar_multiply(alpha * v_modulus / F_modulus, force_per_atom, temp2);
+    CHECK(cudaDeviceSynchronize());
```
Comment (brucefan1983) [inline src/minimize/minimizer_fire.cu:96]: do not need `cudaDeviceSynchronize`

File: src/minimize/minimizer_fire.cu [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+    vector_sum(temp1, temp2, v);
+    CHECK(cudaDeviceSynchronize());
```
Comment (brucefan1983) [inline src/minimize/minimizer_fire.cu:98]: do not need `cudaDeviceSynchronize`

File: src/minimize/minimizer_fire.cu [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+    // dx = v*dt
+    scalar_multiply(dt, v, temp1);
+    CHECK(cudaDeviceSynchronize());
```
Comment (brucefan1983) [inline src/minimize/minimizer_fire.cu:101]: do not need `cudaDeviceSynchronize`

File: src/minimize/minimizer_fire.cu [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+    vector_sum(position_per_atom, temp1, position_per_atom);
+    CHECK(cudaDeviceSynchronize());
```
Comment (brucefan1983) [inline src/minimize/minimizer_fire.cu:103]: do not need `cudaDeviceSynchronize`

File: src/utilities/vector_algo.cu [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+    You should have received a copy of the GNU General Public License
+    along with GPUMD.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "vector_algo.cuh"
+
+__global__ void gpu_multiply(const int size, double a, double* b, double* c)
```
Comment (brucefan1983) [inline src/utilities/vector_algo.cu:18]: Add `static` before `__global__` to restrict the scope of the function.

File: src/utilities/vector_algo.cu [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+{
+  int n = blockDim.x * blockIdx.x + threadIdx.x;
+  if (n < size)
+    c[n] = b[n] * a;
+}
+
+__global__ void gpu_vector_sum(const int size, double* a, double* b, double* c)
```
Comment (brucefan1983) [inline src/utilities/vector_algo.cu:25]: Add `static` before `__global__` to restrict the scope of the function.

File: src/utilities/vector_algo.cu [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+{
+  int n = blockDim.x * blockIdx.x + threadIdx.x;
+  if (n < size)
+    c[n] = a[n] + b[n];
+}
+
+__global__ void gpu_pairwise_product(const int size, double* a, double* b, double* c)
```
Comment (brucefan1983) [inline src/utilities/vector_algo.cu:32]: Add `static` before `__global__` to restrict the scope of the function.

File: src/utilities/vector_algo.cu [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+    c[n] = a[n] * b[n];
+}
+
+void pairwise_product(GPU_Vector<double>& a, GPU_Vector<double>& b, GPU_Vector<double>& c)
+{
+  int size = a.size();
+  gpu_pairwise_product<<<(size - 1) / 128 + 1, 128>>>(size, a.data(), b.data(), c.data());
+}
+
+__global__ void gpu_sum(const int size, double* a, double* result)
```
Comment (brucefan1983) [inline src/utilities/vector_algo.cu:45]: Add `static` before `__global__` to restrict the scope of the function.

File: src/utilities/vector_algo.cu [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+  gpu_sum<<<1, 1024>>>(a.size(), a.data(), result.data());
+  result.copy_to_host(&ret);
+  return ret;
+}
+
+double dot(GPU_Vector<double>& a, GPU_Vector<double>& b)
+{
+  GPU_Vector<double> temp(a.size());
+  pairwise_product(a, b, temp);
+  CHECK(cudaDeviceSynchronize());
```
Comment (brucefan1983) [inline src/utilities/vector_algo.cu:81]: Do not need `cudaDeviceSynchronize`

**Comment by brucefan1983** [comment]
Currently, the functions in `vector_algo` have not been used in other files. They have very short names are will easily confilict with functions in other files. I suggest move these functions to `minimizer_fire` and put them into an anonymous `namespace` or decorate them as `static`.

**Comment by psn417** [comment]
> Currently, the functions in `vector_algo` have not been used in other files. They have very short names and will easily confilict with functions in other files. I suggest move these functions to `minimizer_fire` and put them into an anonymous `namespace` or decorate them as `static`.
> 
> The point is not to pollute the global namespace.

I've moved them to `minimizer_fire.cu` and make them anonymous.

**Pull request merged**